In [7]:
import pandas as pd  # Para manejar los datos en DataFrames
from sklearn.model_selection import train_test_split  # Para dividir datos en entrenamiento/prueba
from sklearn.preprocessing import StandardScaler  # Para normalizar los datos
from sklearn.ensemble import RandomForestClassifier  # Modelo de clasificación
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix  # Métricas de evaluación
from imblearn.over_sampling import SMOTE  # Para balancear las clases con datos sintéticos


file_path = r"C:\Users\Emiliano\Desktop\SS python VSC\DatosFelkipeVadillo\Metadatos Eugenia V1.xlsx"
xls = pd.ExcelFile(file_path)  # Lee el archivo Excel
df = pd.read_excel(xls, sheet_name='cohorte_princesa_corregidaDESC_')

#  Filtrar solo los casos de interés: Term, Preterm, Postterm (excluyendo Abortion)
df_filtered = df[df['desenlace_parto'].isin(['Term', 'Preterm', 'Postterm'])].copy()

# Reasignar etiquetas a la variable objetivo (conversión a números)
df_filtered['desenlace_parto'] = df_filtered['desenlace_parto'].replace({
    'Term': 0,      # 0 → Parto a término
    'Preterm': 1,   # 1 → Parto pretérmino
    'Postterm': 2   # 2 → Parto postérmino
})

# Seleccionar solo características maternas: edad, peso pregestacional y talla
columns_maternal = ['edad_cronologicamujer', 'peso_pregestacional_kg', 'talla_mujer_cm']
df_maternal = df_filtered[columns_maternal + ['desenlace_parto']]

# Manejo de valores faltantes (rellenar con la media de cada columna)
df_maternal = df_maternal.fillna(df_maternal.mean())

# Separar características (X) y la variable objetivo (y)
X_maternal = df_maternal.drop(columns=['desenlace_parto'])  # Variables predictoras
y_maternal = df_maternal['desenlace_parto']  # Variable objetivo (clase a predecir)

# Normalizar características para mejorar el rendimiento del modelo
scaler = StandardScaler()
X_maternal_scaled = scaler.fit_transform(X_maternal)  # Se escalan los datos

# Dividir los datos en conjunto de entrenamiento (80%) y prueba (20%)
X_train_m, X_test_m, y_train_m, y_test_m = train_test_split(
    X_maternal_scaled, y_maternal, test_size=0.2, random_state=42, stratify=y_maternal
)

# Aplicar SMOTE para balancear las clases en el conjunto de entrenamiento
smote = SMOTE(sampling_strategy='auto', random_state=42, k_neighbors=1)  # Genera datos sintéticos
X_train_m_balanced, y_train_m_balanced = smote.fit_resample(X_train_m, y_train_m)

# Entrenar el modelo Random Forest con los datos balanceados
model_maternal_smote = RandomForestClassifier(n_estimators=100, random_state=42, class_weight="balanced")
model_maternal_smote.fit(X_train_m_balanced, y_train_m_balanced)  # Se entrena el modelo

# Realizar predicciones en el conjunto de prueba
y_pred_maternal_smote = model_maternal_smote.predict(X_test_m)

accuracy_maternal_smote = accuracy_score(y_test_m, y_pred_maternal_smote)  # Precisión del modelo
classification_rep_maternal_smote = classification_report(y_test_m, y_pred_maternal_smote)  # Reporte de clasificación
conf_matrix_maternal_smote = confusion_matrix(y_test_m, y_pred_maternal_smote)  # Matriz de confusión

print(f"Precisión del modelo con SMOTE: {accuracy_maternal_smote:.2f}")  # Se imprime la precisión del modelo
print("Reporte de clasificación:\n", classification_rep_maternal_smote)  # Se imprime el reporte detallado
print("Matriz de confusión:\n", conf_matrix_maternal_smote)  # Se imprime la matriz de confusión



Precisión del modelo con SMOTE: 0.88
Reporte de clasificación:
               precision    recall  f1-score   support

           0       0.86      1.00      0.92        18
           1       1.00      0.40      0.57         5
           2       1.00      1.00      1.00         1

    accuracy                           0.88        24
   macro avg       0.95      0.80      0.83        24
weighted avg       0.89      0.88      0.85        24

Matriz de confusión:
 [[18  0  0]
 [ 3  2  0]
 [ 0  0  1]]


C:\Users\Emiliano\AppData\Local\Temp\ipykernel_17580\2360491043.py:17: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_filtered['desenlace_parto'] = df_filtered['desenlace_parto'].replace({


Precisión del modelo (0.85 en este caso)

Indica que el modelo tiene un 88% de precisión en el conjunto de prueba.
Reporte de clasificación

Contiene las métricas clave para cada clase (Term, Preterm, Postterm):
Precision: Cuántas predicciones positivas fueron realmente correctas.
Recall: Cuántos casos de cada clase fueron correctamente identificados.
F1-score: Combinación de Precision y Recall.
Support: Cantidad de casos en cada clase.
Matriz de confusión

Muestra cómo se distribuyen las predicciones entre las clases reales.
Por ejemplo, [[46 2 2] significa que:
46 casos de "Term" fueron correctamente clasificados como "Term".
2 casos de "Term" fueron mal clasificados como "Preterm".
2 casos de "Term" fueron mal clasificados como "Postterm".